packages/drivers/odsp-driver/src/compactSnapshotParser.ts

```diff
@@ -0,0 +1,156 @@
+
+export const snapshotMinReadVersion = "1.0";
+
+/**
+ *  Header of the downloaded snapshot.
+ */
+export interface ISnapshotHeader {
+    // This is the minimum version of the reader required to read the wire format of a given snapshot.
+    minReadVersion: string,
+    // Represents the version with which snapshot is created.
+    createVersion: string,
+    // Seq number at which the snapshot is created.
+    snapshotSeqNumber: number,
+}
+
+/**
+ * Recreates header section and then validate that.
+ * @param node - tree node to read header section from
+ */
+function readAndValidateHeaderSection(node: NodeCore) {
+    let minReadVersion: string | undefined;
+    let createVersion: string | undefined;
+    for (const [headerName, value] of node.iteratePairs()) {
+        assert(headerName instanceof BlobCore, "Header name should be string");
+        switch(headerName.toString()) {
+            case "MinReadVersion": {
+                assert(value instanceof BlobCore, "MinReadVersion should be a string");
+                minReadVersion = value.toString();
+                break;
+            }
+            case "CreateVersion": {
+                assert(value instanceof BlobCore, "CreateVersion should be a string");
+                createVersion = value.toString();
+                break;
+            }
+            default: {
+                break;
+            }
+        }
+    }
+
+    assert(minReadVersion !== undefined &&
+        snapshotMinReadVersion >= minReadVersion, "Driver min read version should >= to server minReadVersion");
+    assert(createVersion !== undefined &&
+        createVersion >= snapshotMinReadVersion, "Snapshot should be created with minReadVersion or above");
+}
+
+/**
+ * Recreates dictionary section of the tree.
+ * @param node - tree node to read dictionary section from. This container paths/ids mapping
+ * to integer representation.
+ */
+function readDictionarySection(node: NodeCore) {
+    const dictionary = new Array<string>();
+    for (const name of node) {
+        assert(name instanceof BlobCore, "Mapping should be of type BlobCore");
+        dictionary.push(name.toString());
+    }
+    return dictionary;
+}
+
+/**
+ * Recreates blobs section of the tree.
+ * @param node - tree node to read blob section from
+ * @param dictionary - name map, used to decode path/IDs.
+ */
+function readBlobSection(node: NodeCore, dictionary: string[]) {
+    const blobs: Map<string, ArrayBuffer> = new Map();
+    for (const [idIndex, blob] of node.iteratePairs()) {
+        assert(typeof idIndex === "number", "Blob index should be a number");
+        assert(blob instanceof BlobCore, "Blob content should be of type blob");
+        const blobId = dictionary[idIndex];
+        assert(blobId !== undefined, "blob id should be present");
+        const unit8Array = blob.buffer;
+        blobs.set(blobId,
+            unit8Array.buffer.slice(unit8Array.byteOffset, unit8Array.byteOffset + unit8Array.byteLength));
```

Comment:
Isn't it equivalent to just using unit8Array? I thought Uint8Array conforms to ArrayBuffer signature, doesn't it?
also unit -> uint?

Reply:
No. unitArray.buffer is specifically of type ArrayBuffer. We can put uint8 in var of type ArrayBuffer but it would fail the "instanceof" check.

Reply:
I'd suggest moving this logic directly into BlobCore, or into stand-alone static function.
I do not see any way to do shallow copy here :(
Which suggests that our usage of ArrayBuffer for driver API may not have been good choice - Uint8Array would have been a better choice. I guess something to think about for future improvements :)